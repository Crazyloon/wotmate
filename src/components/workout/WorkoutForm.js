<<<<<<< HEAD
import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../shared/TextInput';
import SelectInput from '../shared/SelectInput';
import LiftForm from './subforms/LiftForm';
import CardioForm from './subforms/CardioForm';
import ActivityForm from './subforms/ActivityForm';
import { workoutTypes } from '../../data/api/workoutApi';

const WorkoutForm = ({ workout, exercise, newSet, customWorkout, onNameChange,
                       onExerciseChange, onLiftChange, onAddSet, onCardioChange,
                       onActivityChange, onScaleChange, saving, errors,
                       durationOptions, weightOptions, distanceOptions,
                       durationScale, weightScale, distanceScale}) => {
  let formClass = `needs-validation`;
  let dropDownOptions = [{value: "lift", text: "Lift"}, 
                         {value: "cardio", text:"Cardio"}, 
                         {value: "activity", text:"Activity"}];
  // errors.style ? formClass = 'needs-validation was-validated' : null;
  function form(tag){
    switch(tag){
      case 'lift':
        return (
          <LiftForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onLiftChange={onLiftChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            weightOptions={weightOptions}
            weightScale={weightScale}
          />
        );
        case 'cardio':
        return (
          <CardioForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onCardioChange={onCardioChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            durationOptions={durationOptions}
            distanceOptions={distanceOptions}
            durationScale={durationScale}
            distanceScale={distanceScale}
          />
        );
        case 'activity':
        return (
          <ActivityForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onActivityChange={onActivityChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            durationOptions={durationOptions}
            durationScale={durationScale}
          />
        );
    }
  }

  return (
    <div className="col-md-8 offset-md-2">
      <h1>Manage Workout</h1>
      <form className={formClass} noValidate>
        <TextInput 
          wrapperClass={"form-group"}
          name="name"
          label="Workout Name"
          placeholder="Fat Buster"
          value={workout.name}
          onChange={onNameChange}
          error={errors.name}
          />

        <SelectInput
          name="type"
          label="Type"
          onChange={onExerciseChange}
          defaultOption={"Select Workout Type"}
          value={exercise.tag}
          error={errors.type}
          options={dropDownOptions}
          />
        
        <hr style={{height: 2, background: 'white', width: '96%'}}/>

        {form(exercise.tag)}
      </form>
    </div>
  );
};
 
WorkoutForm.propTypes = {
  workout: PropTypes.object.isRequired,
  exercise: PropTypes.object.isRequired,
  newSet: PropTypes.object.isRequired,
  customWorkout: PropTypes.string,
  onNameChange: PropTypes.func.isRequired,
  onExerciseChange: PropTypes.func.isRequired,
  onLiftChange: PropTypes.func.isRequired,
  onCardioChange: PropTypes.func.isRequired,
  onActivityChange: PropTypes.func.isRequired,
  onAddSet: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  errors: PropTypes.object
};

=======
import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../shared/TextInput';
import SelectInput from '../shared/SelectInput';
import LiftForm from './subforms/LiftForm';
import CardioForm from './subforms/CardioForm';
import ActivityForm from './subforms/ActivityForm';
import { workoutTypes } from '../../data/api/workoutApi';

const WorkoutForm = ({ workout, exercise, newSet, customWorkout, onNameChange,
                       onExerciseChange, onLiftChange, onAddSet, onCardioChange,
                       onActivityChange, onScaleChange, saving, errors,
                       durationOptions, weightOptions, distanceOptions,
                       durationScale, weightScale, distanceScale}) => {
  let formClass = `needs-validation`;
  let dropDownOptions = [{value: "lift", text: "Lift"}, 
                         {value: "cardio", text:"Cardio"}, 
                         {value: "activity", text:"Activity"}];
  // errors.style ? formClass = 'needs-validation was-validated' : null;
  function form(tag){
    switch(tag){
      case 'lift':
        return (
          <LiftForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onLiftChange={onLiftChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            weightOptions={weightOptions}
            weightScale={weightScale}
          />
        );
        case 'cardio':
        return (
          <CardioForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onCardioChange={onCardioChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            durationOptions={durationOptions}
            distanceOptions={distanceOptions}
            durationScale={durationScale}
            distanceScale={distanceScale}
          />
        );
        case 'activity':
        return (
          <ActivityForm
            errors={errors}
            workout={workout}
            exercise={exercise}
            newSet={newSet}
            onActivityChange={onActivityChange}
            onAddSet={onAddSet}
            onScaleChange={onScaleChange}
            durationOptions={durationOptions}
            durationScale={durationScale}
          />
        );
    }
  }

  return (
    <div className="col-md-8 offset-md-2">
      <form className={formClass} noValidate>
        <h1>Manage Workout</h1>

        <TextInput 
          wrapperClass={"form-group"}
          name="name"
          label="Workout Name"
          placeholder="Fat Buster"
          value={workout.name}
          onChange={onNameChange}
          error={errors.name}
          />

        <SelectInput
          name="type"
          label="Type"
          onChange={onExerciseChange}
          defaultOption={"Select Workout Type"}
          value={exercise.tag}
          error={errors.type}
          options={dropDownOptions}
          />
        
        {form(exercise.tag)}
      </form>
    </div>
  );
};
 
WorkoutForm.propTypes = {
  workout: PropTypes.object.isRequired,
  exercise: PropTypes.object.isRequired,
  newSet: PropTypes.object.isRequired,
  customWorkout: PropTypes.string,
  onNameChange: PropTypes.func.isRequired,
  onExerciseChange: PropTypes.func.isRequired,
  onLiftChange: PropTypes.func.isRequired,
  onCardioChange: PropTypes.func.isRequired,
  onActivityChange: PropTypes.func.isRequired,
  onAddSet: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  errors: PropTypes.object
};

>>>>>>> fc2641cb0f23a39a3e85159a617b456835d0b532
export default WorkoutForm;