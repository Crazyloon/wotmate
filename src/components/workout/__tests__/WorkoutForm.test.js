import 'babel-polyfill';
import sinon from 'sinon';
import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({
  adapter: new Adapter()
});

import WorkoutForm from '../WorkoutForm'; 

// arrange
const initialData = {
  workout: {
    id: '',
    date: '',
    name: '',
    muscles: [],
    exercises: []
  },
  exercise: {
    tag: '',
    type: {},
    sets: []
  },
  newSet: {
    name: ''
  },
  saving: false,
  errors: {},
  customWorkout: ''
};

const onChangeMock = jest.fn();

const mountedWrapper = mount(<WorkoutForm 
  workout={initialData.workout}
  exercise={initialData.exercise}
  newSet={initialData.newSet}
  customWorkout={initialData.customWorkout} 
  onNameChange={onChangeMock} 
  onExerciseChange={onChangeMock} 
  onLiftChange={onChangeMock} 
  onCardioChange={onChangeMock} 
  onActivityChange={onChangeMock} 
  onAddSet={onChangeMock}
  onSave={onChangeMock} 
  errors={initialData.errors}
  saving={initialData.saving}
/>);

describe('Workout Form', () => {

  describe('initial form state', () =>{
    it('has a name text input', ()=>{
      expect(mountedWrapper.find('TextInput').length).toBe(1);
    });

    it('has a type select input', ()=>{
      expect(mountedWrapper.find('SelectInput').length).toBe(1);
    });

    it('simulates exercise type select correctly', ()=>{
      const onChangeSpy = sinon.spy();
      const shallowWrapper = shallow(<WorkoutForm 
        workout={initialData.workout}
        exercise={initialData.exercise}
        newSet={initialData.newSet}
        customWorkout={initialData.customWorkout} 
        onNameChange={onChangeMock} 
        onExerciseChange={onChangeSpy} 
        onLiftChange={onChangeMock} 
        onCardioChange={onChangeMock} 
        onActivityChange={onChangeMock} 
        onAddSet={onChangeMock}
        onSave={onChangeMock} 
        errors={initialData.errors}
        saving={initialData.saving}
      />);

      const simulatedInput = {target: {value: 'lift'}};
      shallowWrapper.find('SelectInput').first().simulate('change', simulatedInput);      
      sinon.assert.calledWith(onChangeSpy, simulatedInput);
      sinon.assert.calledOnce(onChangeSpy);
      expect(shallowWrapper).toMatchSnapshot();
    });
  });

  describe('selecting "Lift"', ()=>{
    it('presents the LiftForm', ()=>{
      mountedWrapper.setProps({
        exercise: {tag: 'lift', type: {}, sets: []},
        newSet: {
          name: 'Bench Press',
          reps: 8,
          weight: {
            value: 100,
            scale: 'lbs'
          }
        }
      }, () => {mountedWrapper.update();});
    
      expect(mountedWrapper.find('SelectInput').first().props().value).toBe('lift');      
      expect(mountedWrapper.find('LiftForm').length).toBe(1);
      expect(mountedWrapper.find('TextInput').length).toBe(2);
      expect(mountedWrapper.find('SelectInput').length).toBe(2);
      expect(mountedWrapper.find('NumberInput').length).toBe(2);
      expect(mountedWrapper.find('.btn').length).toBe(2);      
    });    
  });

  describe('selecting "Cardio"', ()=>{
    it('presents the CardioForm', ()=>{
      mountedWrapper.setProps({
        exercise: {tag: 'cardio', type: {}, sets: []},
        newSet: {
          name: 'Sprint',
          duration: {
            value: 30,
            scale: 'sec'
          },
          distance: {
            value: 100,
            scale: 'yrds'
          }
        }
      }, () => {mountedWrapper.update();});
     
      expect(mountedWrapper.find('SelectInput').first().props().value).toBe('cardio');      
      expect(mountedWrapper.find('CardioForm').length).toBe(1);
      expect(mountedWrapper.find('TextInput').length).toBe(2);
      expect(mountedWrapper.find('SelectInput').length).toBe(2);
      expect(mountedWrapper.find('NumberInput').length).toBe(2);
      expect(mountedWrapper.find('.btn').length).toBe(2);           
    });    
  });

  describe('selecting "Activity"', ()=>{
    it('presents the ActivityForm', ()=>{
      mountedWrapper.setProps({
        exercise: {tag: 'activity', type: {}, sets: []},
        newSet: {
          name: 'Sprint',
          duration: {
            value: 30,
            scale: 'min'
          },
          intensity: 'Variable'
        }
      }, () => {mountedWrapper.update();});
    
      expect(mountedWrapper.find('SelectInput').first().props().value).toBe('activity');      
      expect(mountedWrapper.find('ActivityForm').length).toBe(1);
      expect(mountedWrapper.find('TextInput').length).toBe(2);
      expect(mountedWrapper.find('SelectInput').length).toBe(3);
      expect(mountedWrapper.find('NumberInput').length).toBe(1);
      expect(mountedWrapper.find('.btn').length).toBe(2);        
    });    
  });
});