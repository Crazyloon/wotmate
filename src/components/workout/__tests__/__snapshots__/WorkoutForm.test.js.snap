// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout Form initial form state simulates exercise type select correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WorkoutForm
    customWorkout=""
    distanceOptions={
      Array [
        "yd",
        "mi",
      ]
    }
    distanceScale="yd"
    durationOptions={
      Array [
        "sec",
        "min",
      ]
    }
    durationScale="sec"
    errors={Object {}}
    exercise={
      Object {
        "sets": Array [],
        "tag": "",
        "type": Object {},
      }
    }
    newSet={
      Object {
        "name": "",
      }
    }
    onActivityChange={[MockFunction]}
    onAddSet={[MockFunction]}
    onCardioChange={[MockFunction]}
    onExerciseChange={[Function]}
    onLiftChange={[MockFunction]}
    onNameChange={[MockFunction]}
    onSave={[MockFunction]}
    onScaleChange={[MockFunction]}
    saving={false}
    weightOptions={
      Array [
        "lb",
        "kg",
      ]
    }
    weightScale="lb"
    workout={
      Object {
        "date": "",
        "exercises": Array [],
        "id": "",
        "muscles": Array [],
        "name": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Manage Workout
        </h1>,
        <form
          className="needs-validation"
          noValidate={true}
        >
          <TextInput
            error={undefined}
            label="Workout Name"
            name="name"
            onChange={[MockFunction]}
            placeholder="Fat Buster"
            value=""
            wrapperClass="form-group"
          />
          <SelectInput
            defaultOption="Select Workout Type"
            error={undefined}
            label="Type"
            name="type"
            onChange={[Function]}
            options={
              Array [
                Object {
                  "text": "Lift",
                  "value": "lift",
                },
                Object {
                  "text": "Cardio",
                  "value": "cardio",
                },
                Object {
                  "text": "Activity",
                  "value": "activity",
                },
              ]
            }
            value=""
            wrapperClass="form-group"
          />
          <hr
            style={
              Object {
                "background": "white",
                "height": 2,
                "width": "96%",
              }
            }
          />
        </form>,
      ],
      "className": "col-md-8 offset-md-2",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Manage Workout",
        },
        "ref": null,
        "rendered": "Manage Workout",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TextInput
              error={undefined}
              label="Workout Name"
              name="name"
              onChange={[MockFunction]}
              placeholder="Fat Buster"
              value=""
              wrapperClass="form-group"
            />,
            <SelectInput
              defaultOption="Select Workout Type"
              error={undefined}
              label="Type"
              name="type"
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "text": "Lift",
                    "value": "lift",
                  },
                  Object {
                    "text": "Cardio",
                    "value": "cardio",
                  },
                  Object {
                    "text": "Activity",
                    "value": "activity",
                  },
                ]
              }
              value=""
              wrapperClass="form-group"
            />,
            <hr
              style={
                Object {
                  "background": "white",
                  "height": 2,
                  "width": "96%",
                }
              }
            />,
            undefined,
          ],
          "className": "needs-validation",
          "noValidate": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "error": undefined,
              "label": "Workout Name",
              "name": "name",
              "onChange": [MockFunction],
              "placeholder": "Fat Buster",
              "value": "",
              "wrapperClass": "form-group",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "defaultOption": "Select Workout Type",
              "error": undefined,
              "label": "Type",
              "name": "type",
              "onChange": [Function],
              "options": Array [
                Object {
                  "text": "Lift",
                  "value": "lift",
                },
                Object {
                  "text": "Cardio",
                  "value": "cardio",
                },
                Object {
                  "text": "Activity",
                  "value": "activity",
                },
              ],
              "value": "",
              "wrapperClass": "form-group",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "style": Object {
                "background": "white",
                "height": 2,
                "width": "96%",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "hr",
          },
          undefined,
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Manage Workout
          </h1>,
          <form
            className="needs-validation"
            noValidate={true}
          >
            <TextInput
              error={undefined}
              label="Workout Name"
              name="name"
              onChange={[MockFunction]}
              placeholder="Fat Buster"
              value=""
              wrapperClass="form-group"
            />
            <SelectInput
              defaultOption="Select Workout Type"
              error={undefined}
              label="Type"
              name="type"
              onChange={[Function]}
              options={
                Array [
                  Object {
                    "text": "Lift",
                    "value": "lift",
                  },
                  Object {
                    "text": "Cardio",
                    "value": "cardio",
                  },
                  Object {
                    "text": "Activity",
                    "value": "activity",
                  },
                ]
              }
              value=""
              wrapperClass="form-group"
            />
            <hr
              style={
                Object {
                  "background": "white",
                  "height": 2,
                  "width": "96%",
                }
              }
            />
          </form>,
        ],
        "className": "col-md-8 offset-md-2",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Manage Workout",
          },
          "ref": null,
          "rendered": "Manage Workout",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TextInput
                error={undefined}
                label="Workout Name"
                name="name"
                onChange={[MockFunction]}
                placeholder="Fat Buster"
                value=""
                wrapperClass="form-group"
              />,
              <SelectInput
                defaultOption="Select Workout Type"
                error={undefined}
                label="Type"
                name="type"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "text": "Lift",
                      "value": "lift",
                    },
                    Object {
                      "text": "Cardio",
                      "value": "cardio",
                    },
                    Object {
                      "text": "Activity",
                      "value": "activity",
                    },
                  ]
                }
                value=""
                wrapperClass="form-group"
              />,
              <hr
                style={
                  Object {
                    "background": "white",
                    "height": 2,
                    "width": "96%",
                  }
                }
              />,
              undefined,
            ],
            "className": "needs-validation",
            "noValidate": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "error": undefined,
                "label": "Workout Name",
                "name": "name",
                "onChange": [MockFunction],
                "placeholder": "Fat Buster",
                "value": "",
                "wrapperClass": "form-group",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "defaultOption": "Select Workout Type",
                "error": undefined,
                "label": "Type",
                "name": "type",
                "onChange": [Function],
                "options": Array [
                  Object {
                    "text": "Lift",
                    "value": "lift",
                  },
                  Object {
                    "text": "Cardio",
                    "value": "cardio",
                  },
                  Object {
                    "text": "Activity",
                    "value": "activity",
                  },
                ],
                "value": "",
                "wrapperClass": "form-group",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "style": Object {
                  "background": "white",
                  "height": 2,
                  "width": "96%",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "hr",
            },
            undefined,
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
